### DISCUSSION TITLE
Start Tiddlywiki on Node.js using random port

### DISCUSSION THREAD

**Discussion by kookma** [TiddlyWiki/TiddlyWiki5#5706] [Ideas]
I have frequently start different Tiddlywiki by visiting the folder and issue the below command

```
tiddlywiki --server 3000
```

or simply

```
tiddlywiki --listen
```

Sometime I forget which ports are available and I get error!!

I am wondering if we could have an option to start Tiddlywiki on a random port or next vacant port like

```
tiddlywiki --listen random
```

or

```
tiddlywiki --server random
```

**Comment by linonetwo** [comment]
Maybe you can use a script to run wiki:

```js
import path from 'path';
import tiddlywiki from 'tiddlywiki';

function startNodeJSWiki({ homePath, tiddlyWikiPort = Math.floor(Math.random() * 13000) - 3000, userName }) {
  return new Promise((resolve, reject) => {
    const wikiInstance = tiddlywiki.TiddlyWiki();
    try {
      process.env.TIDDLYWIKI_PLUGIN_PATH = path.resolve(homePath, 'plugins');
      process.env.TIDDLYWIKI_THEME_PATH = path.resolve(homePath, 'themes');
      // add tiddly filesystem back https://github.com/Jermolene/TiddlyWiki5/issues/4484#issuecomment-596779416
      wikiInstance.boot.argv = [
        '+plugins/tiddlywiki/filesystem',
        '+plugins/tiddlywiki/tiddlyweb',
        homePath,
        '--listen',
        `anon-username=${userName}`,
        `port=${tiddlyWikiPort}`,
        `host=${defaultServerIP}`,
        'root-tiddler=$:/core/save/lazy-images',
      ];
      wikiInstance.boot.startup({
        callback: () => resolve(`Tiddlywiki booted at http://localhost:${tiddlyWikiPort}`),
      });
    } catch (error) {
      const message = `Tiddlywiki booted failed with error ${error.message} ${error.stack ?? ''}`;
      reject(message);
    }
  });
}

```

And maybe check for port usage before use it, and reroll if used.

**Reply by linonetwo** [reply]
I use GUI to manage port for each wiki, so it will never collide.

**Reply by kookma** [reply]
Hi @linonetwo 
Wonderful! Thank you!

I assume it should be in the path (I am on Windows 10) and be called like

```
startNodeJSWiki   mywiki 
```

**Reply by linonetwo** [reply]
I think the JS should be the next Shell script, with https://github.com/google/zx or write it your self (just save my script to a .mjs file) and add it to the `path`.

You can use `process.argv` to get parameters as https://github.com/google/zx/issues/83 said.

**Reply by linonetwo** [reply]
I think probably you can add your `tiddlers` folder to the path, and use tiddlywiki itself to manage scripts? I haven't try this, but I think this will make manage scripts easier, easier to search and find and backup.

**Reply by kookma** [reply]
>  think the JS should be the next Shell script, with https://github.com/google/zx or write it your self (just save my script to a .mjs file) and add it to the path.

Thank you @linonetwo I got the point!

> I think probably you can add your tiddlers folder to the path, and use tiddlywiki itself to manage scripts?

This is very interesting! I have not do this before!
